### PR TITLE
Implement login and logout

### PR DESCRIPTION
**Pull request by tfidfwastaken**
Closes [sc-135].

We store sessions in memory for the time being. Adding a persistent external store is possible without change to this code by implementing Ring's [SessionStore protocol](https://github.com/ring-clojure/ring/wiki/Sessions#session-stores). This is a task reserved for later (cf. [ADR-001](https://github.com/nilenso/bugle-forms/blob/4cf9e36edc5580edc9159e10b3d0678b671dc28c/doc/arch/adr-001-sessions-for-auth.md)).

### CODE REVIEW

File: src/bugle_forms/handlers/user.clj [hunk 1]
```diff
@@ -8,17 +8,33 @@
...
+    (response/redirect "/dashboard" :see-other)
+    (response/response (layout/application
+                        {:title "Sign up" :flash flash}
+                        user-views/signup))))
+
+(defn create-session
+  "Create a session on a successful login."
+  [{:keys [form-params session]}]
+  (if (:user session)
+    (response/redirect "/dashboard" :see-other)
```
Comment (shafeeq) [inline src/bugle_forms/handlers/user.clj]: Use a middleware that can do the redirect to `/dashboard` if the session already exists. That way we can avoid checking for session in each handler.

File: src/bugle_forms/model/user.clj [hunk 2]
```diff
@@ -24,3 +28,27 @@
...
+  Returns a map containing username and a random session ID if successful, or an
+  error value otherwise."
+  [{email :email, given-pwd :password}]
+  (if-let [{pwd :password, name :name}
+           (plan/select-one!
+            db/datasource [:name :password]
+            ["select name, password from user_account where email = ?" email])]
+    (if (:valid (hashers/verify given-pwd pwd))
+      {:id (generate-session-id)
+       :name name}
```
Comment (shafeeq) [inline src/bugle_forms/model/user.clj:53]: We should store some user identifier in session for use later. `name` is not a good identifier becuse it's not unique.

File: src/bugle_forms/handlers/user.clj [hunk 3]
```diff
@@ -8,17 +8,33 @@
...
-                      {:title "Sign up" :flash flash}
-                      user-views/signup)))
+  [{:keys [flash], {:keys [user]} :session}]
+  (if user
+    (response/redirect "/dashboard" :see-other)
+    (response/response (layout/application
+                        {:title "Sign up" :flash flash}
+                        user-views/signup))))
+
+(defn create-session
```
Comment (shafeeq) [inline src/bugle_forms/handlers/user.clj]: The name `create-session` is a bit confusing. It seems to suggest that the fn mostly takes care of creating session based on the current implementation, but here it's actually the handler for the POST request for login, which might or might not create the session based on credential verification.

Maybe rename this to `login` and the other GET handler to `login-form`?

File: test/bugle_forms/handlers/user_test.clj [hunk 4]
```diff
@@ -28,5 +28,79 @@
...
+        (is (:session response))
+        (is (= 303 (:status response)))
+        (is (= {"Location" "/dashboard"} (:headers response)))))
+
+    (testing "Unsuccessful login redirects to login page again"
+      (let [login-form-params {:email (:email signup-form-params)
+                               :password "hunter2"}
+            login-request {:form-params login-form-params}
+            response (sut/create-session login-request)]
+        (is (= 303 (:status response)))
```
Comment (shafeeq) [inline test/bugle_forms/handlers/user_test.clj:54]: Also test for absence of session?

File: test/bugle_forms/model/user_test.clj [hunk 5]
```diff
@@ -15,3 +20,25 @@
...
+(deftest authenticate
+  (let [user-data {:name "Ghanshyam Pinto"
+                   :email "gs@nilen.so"
+                   :password "t0ps3cr3t"}
+        a-user (user/create user-data)
+        _ (user/insert! a-user)]
+    (testing "Existing user is assigned a valid session ID"
+      (let [login-params (select-keys user-data [:email :password])
+            auth-response (user/authenticate login-params)]
+        (is (string? (:id auth-response)))
```
Comment (shafeeq) [inline test/bugle_forms/model/user_test.clj:33]: Also test for the id being not-empty, or a fixed length if it is expected.

**Review by shafeeq — CHANGES_REQUESTED** [review]
Looks good, requested some minor changes

File: src/bugle_forms/handlers/user.clj (hunk 1 above)
Comment (tfidfwastaken) [reply src/bugle_forms/handlers/user.clj]: Makes sense. 

Related: would it make sense to have a middleware for the converse of this—ie, trying to access the dashboard _without_ a session should redirect to `/login`. This case will be more common than the one you are talking about—there will be more privileged routes than unprivileged ones down the line.

Thinking off-the-cuff right now, but I think these two cases should be unified somehow in the same middleware.

File: src/bugle_forms/model/user.clj (hunk 2 above)
Comment (tfidfwastaken) [reply src/bugle_forms/model/user.clj:53]: Okay. I didn't intend to use `:name` as an identifier, rather it was meant as data to personalize messages to a logged in user ("Hi, `:name` 👋 "). I didn't actually think of having including the user's unique identifier at all in the session, but I'll eventually need it for sure. Thanks for calling this out.

File: src/bugle_forms/handlers/user.clj (hunk 1 above)
Comment (shafeeq) [reply src/bugle_forms/handlers/user.clj]: You're right, we'll need to handle redirects to login as well. These two can be different middlewares that we can selectively apply to handlers

File: src/bugle_forms/handlers/user.clj [hunk 6]
```diff
@@ -6,25 +6,48 @@
...
       (-> (response/redirect "/login" :see-other)
           (assoc :flash "Account creation successful!"))
-      (util/error-redirect "/signup" "User already exists. Try logging in."))))
+      (util/flash-redirect "/signup" "User already exists. Try logging in."))))
+
+(defn logout
+  "Log the user out; deletes session."
+  [{{:keys [user]} :session}]
+  (-> (util/flash-redirect "/" "Logged out successfully.")
+      (assoc :session nil)))
```
Comment (shafeeq) [inline src/bugle_forms/handlers/user.clj:46]: Use `dissoc`

File: src/bugle_forms/middleware.clj [hunk 7]
```diff
@@ -24,3 +28,51 @@
...
       (handler request-kw-form-params))))
+
+(defn validate-request
+  "Guard against malformed requests.
+  Takes an map containing a handler and options to validate the request.
+
+  If the `:spec` option is given, the handler is called only when the `:field`
+  value in the request matches the spec, else a 400 status response is
+  returned."
+  [handler {:keys [spec field]}]
```
Comment (shafeeq) [inline src/bugle_forms/middleware.clj:39]: Is field a necessary argument? Can we always just read `params` from the request map?

File: test/bugle_forms/handlers/user_test.clj [hunk 8]
```diff
@@ -28,5 +28,41 @@
...
+        (is (= {"Location" "/dashboard"} (:headers response)))))
+
+    (testing "Unsuccessful login redirects to login page again"
+      (let [login-form-params {:email (:email signup-form-params)
+                               :password "hunter2"}
+            login-request {:form-params login-form-params}
+            response (sut/login login-request)]
+        (is (= 303 (:status response)))
+        (is (nil? (get-in response [:headers "Set-Cookie"])))
```
Comment (shafeeq) [inline test/bugle_forms/handlers/user_test.clj]: Since the request is made directly to the user handler, the response here won't have any headers for setting the session cookie - it is added when it passes through the session middleware in the middleware chain.

At this point, the tests for session should be just based on `:session` in response map

**Review by shafeeq — COMMENTED** [review]
Good job on extracting the middleware usage to a declarative form. The tests for it looks good, but they all use generated handlers. I would suggest adding a test that simulates a request all the way through the middleware chain as well - just to verify that they are integrated correctly.

File: src/bugle_forms/middleware.clj (hunk 7 above)
Comment (tfidfwastaken) [reply src/bugle_forms/middleware.clj:39]: In my opinion, it is useful because we might want to validate our spec on various possible request fields. Eg:
- Validate form submission fields only, ie, `:form-params`
- Validate query strings only, ie, `:query-params`
- Validate some combination of both, ie, `:params`
- (more esoteric?) Validate some other request fields, eg, `:scheme` should be `https` or `X-csrf-token` header exists.

Granted, I only ever validate form-params in my application so far, so this may have been a case of premature abstraction.

File: test/bugle_forms/handlers/user_test.clj (hunk 8 above)
Comment (tfidfwastaken) [reply test/bugle_forms/handlers/user_test.clj]: Got it, thanks for catching this!

File: src/bugle_forms/handlers/user.clj (hunk 6 above)
Comment (tfidfwastaken) [reply src/bugle_forms/handlers/user.clj:46]: This actually won't work in deleting the session. The assoc of nil is information for the Ring `wrap-session` middleware to remove the session. There anyway did not exist a `:session` key in the response map, so there is nothing to `dissoc`, so using that will break logout.
